File: engines/pdf.py
"""PDF Engine — converts a DOCX buffer to PDF via LibreOffice headless."""
import os
import subprocess
import tempfile
import time
from pathlib import Path


class PdfError(Exception):
    pass


def docx_to_pdf(docx_bytes: bytes, timeout_sec: int = 30) -> bytes:
    soffice = os.environ.get('LIBREOFFICE_PATH', 'soffice')
    tmp_dir = Path(tempfile.gettempdir())
    stamp = str(int(time.time() * 1000))
    docx_path = tmp_dir / f'zz-{stamp}.docx'
    pdf_path = tmp_dir / f'zz-{stamp}.pdf'

    try:
        docx_path.write_bytes(docx_bytes)
        result = subprocess.run(
            [soffice, '--headless', '--convert-to', 'pdf', '--outdir', str(tmp_dir), str(docx_path)],
            capture_output=True, text=True, timeout=timeout_sec,
        )
        if not pdf_path.exists():
            raise PdfError(
                f'PDF not created. Is LibreOffice installed / LIBREOFFICE_PATH correct?\n{result.stderr}'
            )
        return pdf_path.read_bytes()
    except subprocess.TimeoutExpired as e:
        raise PdfError(f'PDF conversion timed out after {timeout_sec}s') from e
    finally:
        docx_path.unlink(missing_ok=True)
        pdf_path.unlink(missing_ok=True)

File: web/lib/pdf.ts
/**
 * PDF Engine — faithful TypeScript port of engines/pdf.py
 *
 * Converts a DOCX buffer to PDF via LibreOffice headless.
 * Reads LIBREOFFICE_PATH from environment.
 */
import { execFileSync } from 'child_process';
import fs from 'fs';
import os from 'os';
import path from 'path';

export class PdfError extends Error {}

export function docxToPdf(docxBytes: Buffer, timeoutMs = 60_000): Buffer {
  const soffice = process.env.LIBREOFFICE_PATH ?? 'soffice';
  const tmpDir = os.tmpdir();
  const stamp = Date.now();
  const docxPath = path.join(tmpDir, `zz-${stamp}.docx`);
  const pdfPath = path.join(tmpDir, `zz-${stamp}.pdf`);

  try {
    fs.writeFileSync(docxPath, docxBytes);
    execFileSync(
      soffice,
      ['--headless', '--convert-to', 'pdf', '--outdir', tmpDir, docxPath],
      { timeout: timeoutMs, stdio: 'pipe' }
    );
    if (!fs.existsSync(pdfPath)) {
      throw new PdfError(
        'PDF not created. Is LibreOffice installed and LIBREOFFICE_PATH correct?'
      );
    }
    return fs.readFileSync(pdfPath);
  } catch (e) {
    if (e instanceof PdfError) throw e;
    throw new PdfError(`PDF conversion failed: ${e}`);
  } finally {
    if (fs.existsSync(docxPath)) fs.unlinkSync(docxPath);
    if (fs.existsSync(pdfPath)) fs.unlinkSync(pdfPath);
  }
}
